Bump client's eslint; fix new linting errors

drewrey opened the pull request:
Bump client's eslint; fix new linting errors

- Avoid `object` as a PropType
- Specify a default PropType for non-required props
- Make other props required
- Ignore no-array-index-key until poem lines have DB ids.
- Remove parens from one-argument arrow functions
- Add dangling commas except in functions.
- Remove un-used session from SessionSortablePoemList
- Bump eslint-related dev-dependencies
- Update yarn.lock

michelmansour commented on client/src/Poem.jsx:
```diff
@@ -10,11 +12,16 @@ PoemLine.propTypes = {
-    return <PoemLine line={(line.trim().length === 0) ? NBSP : line} key={index} />;
-  });
+  const poemLineNodes = poem.text.split(/\n/).map(line =>
+    <PoemLine line={(line.trim().length === 0) ? NBSP : line} key={_.uniqueId()} />,
```
Why did this change from using the index as the key to using `underscore.uniqueId()`?

drewrey replied:
This was inspired by [the rule](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-array-index-key.md) "Prevent usage of Array index in keys".

Since each line doesn't have a natural id (vs., e.g., a poem), it made sense to attach some random id, and I found this as a suggestion for such. Perhaps there's a more natural way to accomplish this?

michelmansour replied:
I'm fine with this rule in principle. I'm not a fan of its application here, though, because this array is effectively immutable -- the order isn't going to change. Ultimately, I think we're going to be storing poem lines as entities in their own right, so they will have database IDs that we can use. For now, I'm hesitant to bring in underscore for this. Looking into whether the rule has any flexibility we can configure.

michelmansour replied:
Actually, let's just turn this off for now and revisit it later. In the mean time, let's be vigilant about the values we use for react keys.

drewrey replied:
Okay, I disabled this rule for this line and added a comment about the intention to have IDs in a database for these lines.

I removed underscore as a dependency.

michelmansour commented on client/src/Login.jsx:
```diff
@@ -65,7 +61,19 @@ class Login extends React.Component {
 
 Login.propTypes = {
-  location: PropTypes.object,
+  location: PropTypes.objectOf({
```
👍

michelmansour commented on client/src/index.jsx:
```diff
@@ -17,5 +17,5 @@ render((
     <Route path="/logout" component={Logout} onEnter={requireAuth} />
   </Router>
-  ), document.getElementById('app')
+  ), document.getElementById('app'),
```
Not sure how I feel about this comma. What rule is enforcing this?

drewrey replied:
This is from `comma-dangle` for functions, as [here](http://eslint.org/docs/rules/comma-dangle#functions).

drewrey replied:
Adding:

> "comma-dangle: ["error", {
>   "function": "ignore",
>  }]

to `client/.eslintrc` results in 42 errors.

michelmansour replied:
Could we override this rule to disallow dangling commas for functions:

```json
"comma-dangle": ["error", {
    "function": "never",
}]
```
? What do you think?

drewrey replied:
Okay, I updated the rule to specify `never` for `functions` and `always-multiline` for all other types.

drewrey commented:
Weird, I can't respond to your review.

I re-based off master, updated the yarn.lock file, and made these updates to the sortable sessions for eslint:
- found an un-used prop in `SessionSortablePoemList`
- made a few props required
- removed a few parens around single-var arrow functions.
